certbot/compat/filesystem.py

```diff
@@ -0,0 +1,160 @@
+        },
+        'all': {
+            'read': mode & stat.S_IROTH,
+            'write': mode & stat.S_IWOTH,
+            'execute': mode & stat.S_IXOTH,
+        },
+    }
+
+
+def _generate_windows_flags(rights_desc):
+    # Some notes about how each POSIX right is interpreted.
+    #
+    # For the rights read and execute, we have a pretty bijective relation between
+    # POSIX flags and their generic counterparts on Windows, so we use them directly
+    # (respectively ntsecuritycon.FILE_GENERIC_READ and ntsecuritycon.FILE_GENERIC_EXECUTE).
+    #
+    # But ntsecuritycon.FILE_GENERIC_WRITE does not correspond to what one could expect from a
+    # write access on Linux: for Windows, FILE_GENERIC_WRITE does not include delete, move or
+    # rename. This is something that requires ntsecuritycon.FILE_ALL_ACCESS.
+    # So to reproduce the write right as POSIX, we will apply ntsecuritycon.FILE_ALL_ACCESS
+    # substracted of the rights corresponding to POSIX read and POSIX execute.
+    #
+    # Finally, having read + write + execute gives a ntsecuritycon.FILE_ALL_ACCESS,
+    # so a "Full Control" on the file.
+    #
+    # A complete list of the rights defined on NTFS can be found here:
+    # https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc783530(v=ws.10)#permissions-for-files-and-folders
+    flag = 0
+    if rights_desc['read']:
+        flag = flag | ntsecuritycon.FILE_GENERIC_READ
+    if rights_desc['write']:
+        flag = flag | (ntsecuritycon.FILE_ALL_ACCESS
+                       ^ ntsecuritycon.FILE_GENERIC_READ
+                       ^ ntsecuritycon.FILE_GENERIC_EXECUTE
+                       # Despite bit `512` being present in ntsecuritycon.FILE_ALL_ACCESS, it is
+                       # not effectively applied to the file or the directory.
+                       # As _generate_windows_flags is also used to compare two dacls, we remove
+                       # it right now to have flags that contain only the bits effectively applied
+                       # by Windows.
+                       ^ 512)
+    if rights_desc['execute']:
+        flag = flag | ntsecuritycon.FILE_GENERIC_EXECUTE
+
+    return flag
+
+
+def _compare_dacls(dacl1, dacl2):
+    aces1 = [dacl1.GetAce(index) for index in range(0, dacl1.GetAceCount())]
+    aces2 = [dacl2.GetAce(index) for index in range(0, dacl2.GetAceCount())]
+
+     # Convert PySIDs into hashable objects
+    aces1_refined = []
+    aces2_refined = []
+    for ace in aces1:
+        if len(ace) == 3:
+            aces1_refined.append((ace[0], ace[1], str(ace[2])))
+        else:
+            aces1_refined.append((ace[0], ace[1], ace[2], ace[3], str(ace[4])))  # type: ignore
+    for index, ace in enumerate(aces2):
+        if len(ace) == 3:
+            aces2_refined.append((ace[0], ace[1], str(ace[2])))
+        else:
+            aces2_refined.append((ace[0], ace[1], ace[2], ace[3], str(ace[4])))  # type: ignore
```

Comment:
What is going on here? Why do we always cast the last ACE to a string?

Reply:
Here we are trying to construct sets from the iterables exposed by `GetAce()` that contains tupple. This way, we can leverage the comparison feature of sets, and of tupple, to quickly assert about the equality of everything here, using the appropriate equality operation of each object.

Note that ACE can be defined with three or five elements.

However, in tuple the equal operation can be transfered in each member if these members are hashable. But the last element (third of fith), that is the SID on which the ACE applies, is not hashable. So I convert them in string that are hashable. The string conversion on a PySID is safe because it uses internally `win32security.ConvertSidToStringSid`.

Reply:
Got it! Can you add a comment explaining this for future devs?

Reply:
I also realized that passing by a set was not checking the order. I compare directly the list instead. As do need anymore hashable objects to compare lists, it dramatically simplifies the implementation.